### PR TITLE
[IMP] account: automatically use year range sequence format

### PR DESCRIPTION
**Pull request by rdeodoo**
If the fiscal year "last day" setting is not the 31th of december, it
means that the fiscal year is actually staggered on 2 years.
In this case, we don't want the moves prefix to be `INV/2024/XXX` but
`INV/2023-2024/XXX`.

Such fiscal year ranges are used by default in some countries like in
India, Australia..

Note that the `year_range` support was introduced with commit [1].

[1]: b485e4e

task-3653456

### CODE REVIEW

**Comment by JulienVR** [comment]
I'll review this when it is greened functionally :)

**Comment by rdeodoo** [comment]
@JulienVR Follow your previous message, know that it just got approved functionally :)

File: addons/account/models/account_move.py
Comment (JulienVR) [inline addons/account/models/account_move.py:3099]: Having only 9999 possible documents may be too few, I guess, but one can always resequence anyway :man_shrugging:

File: addons/account/models/account_move.py
Comment (JulienVR) [inline addons/account/models/account_move.py:3114]: well actually its the date_from, date_to of the fiscal year, it's a bit misleading

File: addons/account/wizard/account_resequence.py
Comment (JulienVR) [inline addons/account/wizard/account_resequence.py:127]: not your fault but it's misleading, these are dates, not years

**Review by JulienVR — COMMENTED** [review]
I still need to have a look at the tests and `_get_sequence_date_range` but I'm discovering this part of accounting today :grimacing:.

Thanks for putting so much effort into documenting your code and your commit messages !

I'd just have one remark about your last commit: I really struggled to understand what you were trying to do, i.e.:
1. Introducing `year_range_month`
2. modifying `_get_starting_sequence` to use the `year_range` by default for companies where the fiscal end period is not in December or if the last day is not the 31st: that is, if they use staggered periods

File: addons/account/models/account_move.py
Comment (rdeodoo) [reply addons/account/models/account_move.py:3114]: I adapted to go with 2 explicit variables :+1:

File: addons/account/models/account_move.py
Comment (rdeodoo) [reply addons/account/models/account_move.py]: What I am sure is that we have to keep this commit/table, without this it's just an nightmare to figure what's going on.
But if you prefer we could move that table directly above the regex declaration (where they are defined in `sequence_mixin.py` and here just change the first line of the comment to

```
            # Some regex are catching more sequence formats than we want, so we
            # need to exclude them, see the comment in `sequence_mixin.py` to have
            # a visual hint about it (there is a table explaining it)
```

Let me know what you prefer, I am fine either ways

File: addons/account/models/sequence_mixin.py
Comment (rdeodoo) [reply addons/account/models/sequence_mixin.py:178]: Indeed

File: addons/account/models/sequence_mixin.py
Comment (rdeodoo) [reply addons/account/models/sequence_mixin.py:321]: I tried to be as clear as possible in the commit message:
```
- Replace `date_start.month` by `self[self._sequence_date_field].month`,
  which seems more logical when reading the code -> We want to use the
  document month, not the fiscal year start month. It is actually the
  same value in all sequence cases (monthly, year, year range...) in the
  current code, but next commit will add a "year range monthly" case
  which will need this change.
```

I'll try to explain it better here, with my mind not being so fresh about it since it's bit quite some time.
This change basically does nothing at this point (before the next commits), it's the same "operation".

`self._sequence_date_field` = `date` (which is then using the `date` field of the `account.move` for instance).

First, when reading the code when starting this PR, this line bugged me out and I was wondering why were we using the sequence start month to name the account move rather than using the account move date month.

I dig into the code and figured it was always the same at the end, as the `month` format value is only used by the `_sequence_monthly_regex`, for which the sequence is only one month, and the account move date will always be in the same month as the sequence month obviously, so at the end using the month from the sequence or the account move date will be the same "by chance".

That explained why this line which looked wrong was actually not leading to any issue.

I kept it as is once I figured, but then I actually had to "fix" it because it wouldn't work "by chance" anymore with the new "year range monthly" format which is also now using the `month` format value but which should use the month of the account move, not the first month of the sequence, as the sequence is yearly (despite showing months too), it would always use the first month of the sequence (eg april) instead of the correct month of the account move.

I spent time figuring all this as I never opened any `account` code so every single line of code and concept were new to me, but hopefully it helps understanding more than what I wrote in the commit msg

**Review by rdeodoo — COMMENTED** [review]
Thanks for the review, I just [push forced](https://github.com/odoo/odoo/compare/22cf3528d6f340190cf80e92cf37316d2913249a..8cec5fa8ec8a192878c6ba486a06b95980cb8540) the changes.

> but I'm discovering this part of accounting today

I discovered the whole accounting app/code with this task so I feel you :D 

> I'd just have one remark about your last commit: I really struggled to understand what you were trying to do, i.e.:

Do you mean you struggled to understand the commit msg before reading the code or reading the code?

File: addons/account/models/account_move.py
Comment (JulienVR) [reply addons/account/models/account_move.py]: let's keep it there, hopefully the regex won't change that often :)

File: addons/account/models/sequence_mixin.py
Comment (JulienVR) [reply addons/account/models/sequence_mixin.py:321]: Thanks a lot for the extensive explanations, I got it now... :+1:

File: addons/account/models/account_move.py
Comment (JulienVR) [inline addons/account/models/account_move.py:3134]: typo ? should be the date.day I guess

File: addons/account/models/account_move.py
Comment (JulienVR) [inline addons/account/models/account_move.py:3132]: tricky corner case, did you try it ? I got this ValidationError with a fiscal year ending the 15th of March: 
image

EDIT: found the issue, `fiscalyear_last_month` is a string (in my case: `'3'`) and `self.date.month` is an int (in my case: `3`) so we ignore your subsequent code handling it

**Review by JulienVR — COMMENTED** [review]
Thanks a lot for the explanations !
There's also something that @william-andre raised when I discussed the PR with him: for a sequence "year_range_month" (with end date = 31st of March), the first sequence would be `23-24/03/0001` while the last would be `23-24/01/0001`, so the ordering would put the Jan, Feb and March invoices before the others :/
Was the PO aware ?

File: addons/account/models/account_move.py
Comment (rdeodoo) [reply addons/account/models/account_move.py]: Fixed, indeed

File: addons/account/models/account_move.py
Comment (rdeodoo) [reply addons/account/models/account_move.py:3132]: Indeed, good catch.
I fixed it and tried with the vendor bills, everything seems to be behaving as expected, eg if creating a second bill for second half of april (in case last day is 15 april), it will consider the existing one and increment to 0002, same for first part of april in case there is already one for that first half of the month

**Comment by rdeodoo** [comment]
I've asked @tsb-odoo about your concern, I'll let you know the answer.

**Review by JulienVR — APPROVED** [review]
LGTM now :)
While we wait for a PO answer, maybe it would be nice to have a test that:
* sets a fiscal end date to the 15th of March, for instance
* creates a move and check its sequence for the 1st of March
* does the same for the 20th of march (this is not tested)

**Comment by rdeodoo** [comment]
I [wrote](https://github.com/odoo/odoo/compare/f428d9c15a025c02400cfaf413ef7957520c1131..b979db5a0b1762d8c962c7e166f2c941b68b132c) a test that should cover the staggered year case when the ending/starting date is in the middle of the month.
It's testing that creating multiple moves in either of the half of the month behaves correctly.
Hopefully it will be green on runbot, locally it works fine.

FYI, the test is also covering https://github.com/odoo/odoo/pull/167744#discussion_r1699838110 and https://github.com/odoo/odoo/pull/167744#discussion_r1699857936. The test is failing if you restore any of those 2 mistakes.

**Review by william-andre — COMMENTED** [review]
We could already extract the part that automatically sets the year-range setting when the user uses fiscal years.

The part about the monthly year range is more difficult from a functional PoV, and it is why we didn't do it when adding the year range support:
* right now the month param follows the calendar year while the year follows the fiscal year
* another solution would be to have `01` be the first month of the fiscal year and `12` the last month.

The second solution would allow sorting by name and keep the logical ordering by date, which is usually very valued by accountants.
On the other hand it is harder to implement and might be more mental gymnastic?

**Comment by rdeodoo** [comment]
Note that I've notified the PO about your concern and suggestions, he confirmed with the India team (the ones mainly using this) that it's indeed a problem that would be nice solving.

On my side, if they have been using it like this, I don't know how bad it would be if it was kept ordered by name (and so not totally "correct"), at least in a v1.
I guess tweaking the order name would be functionally the best approach.

I'll wait for the PO answer before moving forward (I'll update about it here).

**Comment by JulienVR** [comment]
Hello @rdeodoo,

I got an answer from @tsb-odoo. He asked the indian POs and they told him they expect the month digits to be the calendar ones (e.g. april = "04") but we wasn't able to confirm with other softwares. The issue is that once it is done, it won't be possible to reverse (otherwise there may have clashes between the old sequence using the calendar month VS a potential new sequence using the fiscal month).

When reading the task specs, it seems the first purpose was to use the year range sequence format when the fiscal end period is different from 31st of December.

Following @william-andre advice, I'd suggest to separate the changes that introduce the `year_range_month` so we can think about it a bit more and take the proper decision, and merge the rest now as it is due for the community days.

**Comment by rdeodoo** [comment]
> Following @william-andre advice, I'd suggest to separate the changes that introduce the year_range_month so we can think about it a bit more and take the proper decision, and merge the rest now as it is due for the community days.

To be sure to understand, are you suggesting that the vendor bills (any documents not being a sales document actually) stop using the monthly format when there is a staggered year?
It's why it was needed, since it was decided that Bill = Monthly and Invoices = Yearly, when introducing the staggered year format, Bill naturally became `Year Range Monthly` (`year_range_month`).

This PR "just" replaced the `2024` by `23-24` eg, but kept the bill month "as is".

As per my commit about adding comments in the code states:
```
- Add comment to explain why some type of documents seem to randomly
  receive an annual sequence while others go through the monthly
  pattern. This is actually an arbitrary decision made with commit [5]
  based on some user feedbacks

[5]: https://github.com/odoo/odoo/commit/10db78c7a5143eff20963cc399e49f5dbee800b2
```

It means we will be breaking this rule when fiscal year is staggered. Is that ok?

**Comment by JulienVR** [comment]
> > Following @william-andre advice, I'd suggest to separate the changes that introduce the year_range_month so we can think about it a bit more and take the proper decision, and merge the rest now as it is due for the community days.
> 
> To be sure to understand, are you suggesting that the vendor bills (any documents not being a sales document actually) stop using the monthly format when there is a staggered year? It's why it was needed, since it was decided that Bill = Monthly and Invoices = Yearly, when introducing the staggered year format, Bill naturally became `Year Range Monthly` (`year_range_month`).
> 
> This PR "just" replaced the `2024` by `23-24` eg, but kept the bill month "as is".
> 
> As per my commit about adding comments in the code states:
> 
> ```
> - Add comment to explain why some type of documents seem to randomly
>   receive an annual sequence while others go through the monthly
>   pattern. This is actually an arbitrary decision made with commit [5]
>   based on some user feedbacks
> 
> [5]: https://github.com/odoo/odoo/commit/10db78c7a5143eff20963cc399e49f5dbee800b2
> ```
> 
> It means we will be breaking this rule when fiscal year is staggered. Is that ok?

Argh... You're right :melting_face:

I don't know what to do honestly. Merging in master would make the PO happy for the community days, but we would have until September 17th (18.0 should be frozen that week) to modify the code. After that, it will not be possible (or very hard).

**Comment by rdeodoo** [comment]
We had a quick discussion with @tsb-odoo this morning and it seems like what is pretty sure is:
- We want to keep the monthly notation for the non-sales (eg bills) documents (as explained on https://github.com/odoo/odoo/pull/167744#issuecomment-2271473500), and so keep the `year_range_month`
- We want to keep 01 for January, we don't want 01 to be April if the fiscal year starts in April

_Note: I'll wait for the final decision (by POs?) on what should exactly be done here before rebasing and checking the 2 pending review comments on the code._

**Comment by JulienVR** [comment]
Well, it's pretty clear then. I think I'll r+ this unless @william-andre doesn't agree ?

**Comment by william-andre** [comment]
As long as the POs understood that we can't go back (they should by now), I'm good.
So I'm good! (didn't review in detail though)

**Comment by JulienVR** [comment]
hey @rdeodoo could you just rebase so I can r+ please ? :blush:

File: addons/account/models/sequence_mixin.py
Comment (rdeodoo) [reply addons/account/models/sequence_mixin.py:31]: Exactly, but as I did not want to go YOLO here and did not want to modify existing code for no good reason (as I don't know all context of those lines, and since the commit explicitly only modified one of the `seq` occurrences and not the others), I kept this commit as a kind of no diff commit, it only moves stuff around but does not change anything.

It's why I left a commit about this exact line in the commit msg:
```
- The `seq` part is the same everywhere except for the `fixed` case.
  It was introduced with commit [2] to not catch too long sequence to
  avoid integer too long error in postresql.
  It's ensure why it was not done in other regexes as they can still
  lead to the same error, on top of not catching the sequence the same
  way. Considering format like:
  1. INV/123456789123456789
  2. INV/2021/123456789123456789
  Why would the 1. case (fixed regex) consider the sequence as
  "123456789" (9 first digits) while case 2. (yearly regex) would
  consider the sequence as "123456789123456789" (all digits)

[1]: https://github.com/odoo/odoo/commit/1a67344694ac648af0e03df9ed6b93338a20e5b7
[2]: https://github.com/odoo/odoo/commit/49afd2f80784302ad0996741ed2c4ab57d3ac54b
```

I'll let you tell me if you really want this (not really related)) functional / business change to be part of this PR :+1: 
If so, I'll do that in a separate commit for clarity.

File: addons/account/models/sequence_mixin.py
Comment (rdeodoo) [reply addons/account/models/sequence_mixin.py]: Why not :D 
Applied in one of the first commit

**Comment by rdeodoo** [comment]
@JulienVR Done
Only thing left right now is this review [comment](https://github.com/odoo/odoo/pull/167744#discussion_r1713389144) but seems unrelated to this PR. I'd rather do that in another PR in case it creates an issue or something (it was explicitly not done in the initial commit).

File: addons/account/models/sequence_mixin.py
Comment (william-andre) [reply addons/account/models/sequence_mixin.py:31]: Ideally it should be done as a bugfix in stable :+1:

Comment (rdeodoo) [reply addons/account/models/sequence_mixin.py:31]: I'll open a PR for that and ping you, perfect :+1:

**Review by JulienVR — APPROVED** [review]
Thanks a lot for your work @rdeodoo :)
@robodoo r+

**Comment by JulienVR** [comment]
@robodoo rebase-ff

**Comment by Feyensv** [comment]
Seems to be the cause of the failing stagings (https://runbot.odoo.com/runbot/build/66802877)
Can you please take a look and r- if needed to avoid more wasted stagings ? Thanks

**Comment by rdeodoo** [comment]
@robodoo r-

**Comment by rdeodoo** [comment]
Clearly the cause indeed, I'll make sure that this standalone build succeed on the dev branch before asking for r+ :grimacing:

**Comment by rdeodoo** [comment]
@JulienVR Tests have been fixed, which comes with a related [enterprise PR](https://github.com/odoo/enterprise/pull/68354) now.
I started the `L10n standalone test` build and I'll let you know when it's green.

**Comment by JulienVR** [comment]
@robodoo r+

**Pull request closed without merging**
